libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/ir/IrOperationsBuilder.kt

```diff
@@ -316,7 +318,7 @@ internal class IrOperationsBuilder(
   private fun GQLFragmentDefinition.toIr(): IrFragmentDefinition {
     val typeDefinition = schema.typeDefinition(typeCondition.name)
 
-    val inferredVariables = inferVariables(schema, allFragmentDefinitions, fieldsOnDisjointTypesMustMerge)
+    val inferredVariables = inferFragmentVariables(this)
```

Comment:
I haven't benchmarked this but this should save a non-trivial amount of work by collecting inferred variables during initial validation.